### DISCUSSION TITLE
How to correctly remove Flow types in a React Native Web project?

### DISCUSSION THREAD

**Discussion by chichke** [babel/babel#11694] [Q&A] [answered]
## Bug Report

**Current behavior**
I was trying to startup a [react-native-web](https://github.com/necolas/react-native-web) project and when i try to add [React Navigation 5](https://reactnavigation.org/docs/web-support/) 
I did try every solution on the web and nothing did work for me...

```
./node_modules/react-native-gesture-handler/DrawerLayout.js
SyntaxError: /Users/chichke/Documents/Work/XXX/node_modules/react-native-gesture-handler/DrawerLayout.js: Support for the experimental syntax 'flow' isn't currently enabled (30:8):

  28 | const SETTLING = 'Settling';
  29 | 
> 30 | export type PropType = {
     |        ^
  31 |   children: any,
  32 |   drawerBackgroundColor?: string,
  33 |   drawerPosition: 'left' | 'right',

Add @babel/plugin-transform-flow-strip-types (https://git.io/vb49g) to the 'plugins' section of your Babel config to enable transformation.
```

**Expected behavior**
Building 😆 

**Babel Configuration**
- Filename: `babel.config.js`

```js
const path = require('path')

module.exports = (api) => {
  api.cache(true);

  return {
    rules: [
      {
        test: path.join(__dirname, "."),
        exclude: /(node_modules)/,
        loader: "babel-loader",
        options: {
          presets: [
            "@babel/preset-env",
            "@babel/react",
            {
              plugins: ["@babel/plugin-proposal-class-properties", "@babel/plugin-transform-flow-strip-types"],
            },
          ],
        },
      },
    ],

    resolve: {
      alias: {
        "react-native$": "react-native-web",
      },
    },
    presets: [
      "react-native-web",
      "module:metro-react-native-babel-preset",
      "@babel/preset-flow",
    ],
    plugins: [
      [
        "@babel/plugin-transform-flow-strip-types",
        {
          loose: true,
        },
      ],
      ["@babel/plugin-proposal-export-default-from"],
      ["@babel/plugin-transform-runtime"],
      [
        "@babel/plugin-proposal-class-properties",
        {
          loose: true,
        },
      ],
      [
        ("module-resolver",
        {
          alias: {
            "^react-native$": "react-native-web",
          },
        }),
      ],
    ].map(require.resolve),
  };
};

```

⚠️  **I have a `webpack.config.js` with same config as babel.config.js**

**Environment**
```
  System:
    OS: macOS 10.15.4
  Binaries:
    Node: 12.14.1 - ~/.nvm/versions/node/v12.14.1/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.13.4 - ~/.nvm/versions/node/v12.14.1/bin/npm
  npmPackages:
    @babel/cli: ^7.10.1 => 7.10.1 
    @babel/core: ^7.10.2 => 7.10.2 
    @babel/plugin-proposal-class-properties: ^7.10.1 => 7.10.1 
    @babel/plugin-proposal-export-default-from: ^7.10.1 => 7.10.1 
    @babel/plugin-transform-flow-strip-types: ^7.10.1 => 7.10.1 
    @babel/plugin-transform-runtime: ^7.10.1 => 7.10.1 
    @babel/preset-flow: ^7.10.1 => 7.10.1 
    babel-plugin-module-resolver: ^4.0.0 => 4.0.0 
```
- Babel version(s): v7.10.1
- Node/npm version: Node:v12.14.1
- OS: OSX 10.15.4,
- Monorepo: yes
- How you are using Babel: loader

**Package.Json**

```
{
  "name": "XXX",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@babel/core": "^7.10.2",
    "@react-native-community/masked-view": "^0.1.10",
    "@react-navigation/core": "^5.10.0",
    "@react-navigation/native": "^5.5.1",
    "@react-navigation/web": "^1.0.0-alpha.9",
    "babel-plugin-module-resolver": "^4.0.0",
    "react": "16.11.0",
    "react-dom": "16.11.0",
    "react-native-gesture-handler": "^1.6.1",
    "react-native-reanimated": "^1.9.0",
    "react-native-safe-area-context": "^3.0.2",
    "react-native-screens": "^2.8.0",
    "react-native-web": "^0.12.3",
    "react-scripts": "3.4.1",
    "typescript": "^3.9.5"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@babel/cli": "^7.10.1",
    "@babel/plugin-proposal-class-properties": "^7.10.1",
    "@babel/plugin-proposal-export-default-from": "^7.10.1",
    "@babel/plugin-transform-flow-strip-types": "^7.10.1",
    "@babel/plugin-transform-runtime": "^7.10.1",
    "@babel/preset-flow": "^7.10.1"
  },
  "jest": {
    "preset": "react-native-web",
    "transform": {
      "^.+\\.js$": "<rootDir>/node_modules/react-native/jest/preprocessor.js"
    }
  }
}

```

- Anything would be greatly appreciated! ❤️

**Comment by existentialism** [comment]
@chichke your babel config doesn't look right:

```
        options: {
          presets: [
            "@babel/preset-env",
            "@babel/react",
            {
              plugins: ["@babel/plugin-proposal-class-properties", "@babel/plugin-transform-flow-strip-types"],
            },
          ],
        },
```

`plugins` should be the same level as `presets`, not an item within it.

**Comment by JLHwung** [comment]
It seems to me that your project is created from `create-react-app` but you are using react native library. 

As is recommended on https://github.com/facebook/create-react-app#react-native, please use `Expo CLI` to initialize your project.

As @existentialism mentioned above, there are also issues in your config file. But I think if you never run `npm run eject`, these config is no-op.

TL;DR

The `react-native-gesture-handler` is in the `node_modules`, which means it is part of dependencies instead of the app code. `react-scripts` will apply an internal preset called
[`babel-preset-react-app/dependencies`](https://github.com/facebook/create-react-app/blob/fa648daca1dedd97aec4fa3bae8752c4dcf37e6f/packages/babel-preset-react-app/dependencies.js) to your dependencies. This preset does not include `@babel/plugin-transform-flow-strip-types` so error is thrown.

**Comment by arelra** [comment, marked as answer]
@chichke  To expand on @existentialism and @JLHwung 

This looks like the result of not transpiling your `react-native` dependencies with the right set of presets/plugins.

When creating a cross platform `react-native` app using `react-native-web` you need to transpile any `react-native` dependencies you may have in `node-modules`.

This is because `react-native` dependencies are not usually published as transpiled or isomorphic browser-friendly versions as normally they'd be piped by `react-native` through [metro-react-native-babel-preset](https://github.com/facebook/metro/blob/master/packages/metro-react-native-babel-preset/package.json) to perform any transpilation necessary to run on iOS or Android.

However `metro-react-native-babel-preset` is a very heavy preset and will create very large bundle sizes for the web.

So when targeting the web you need to make sure that you:
1. Add a `babel-loader` loader to your webpack config for any `react-native` dependencies you may be using
1. Add any relevant presets or plugins required by those dependencies

If you used `create-react-app` to create your web application (and have ejected it) then add the following presets to the options of each `react-native` dependency's `babel-loader`:

1. `babel-preset-react-app` which adds in the majority of common plugins you'll need
2. `babel-preset-react-app/dependencies` which forces Babel to detect any non-ES6 modules

So something like this:

```
  {
    test: /(@?react-(navigation|native)).*\.(ts|js)x?$/,
    exclude: [/react-native-web/, /\.(native|ios|android)\.(ts|js)x?$/],
    loader: 'babel-loader',
    options: {
      presets: [
        [
          require.resolve('babel-preset-react-app/dependencies'),
          { helpers: true },
        ],
        ['babel-preset-react-app'],
      ],
      ...
```

If this is a new error and your webpack config was working previously it's also possible that your loader test regexes are not being hit correctly so you might want to take a look at those.

**Reply by nicolo-ribaudo** [reply]
Thanks for the detailed answer! I have converted this issue to a Discussion so that it's easier to find in the future.

**Reply by ZahraMirali** [reply]
I have the same issue but I did not understand @arelra answer. can you please post the full code of babel and webpack config?

**Reply by arelra** [reply]
HI @ZahraMirali,

Sorry if it wasn't clear, hopefully I can explain in a simpler way:

1. React native dependencies are not usually published as browser-friendly versions - they use all sorts of syntax that browsers don't support ie flow types, ES7+ etc.
1.  React native dependencies normally expect to be run through the preset `metro-react-native-babel-preset` which will transpile the code to run in the JS engine of a mobile device.
1. However this is a problem when targeting a web version of your application using `react-native-web` as it needs to have the dependencies run through some babel presets to transpile the code into something that can run in a browser
1. To fix this for web you need to tell webpack to specifically transpile your react-native dependencies through babel with the appropriate presets (normally webpack default configurations will exclude any dependencies in `node-modules`)
1. The example above transpiles all dependencies named `react-navigation*` or `react-native*` through the preset `babel-preset-react-app` which is a good preset to use for the reasons outline above

If you're new to web development there's a lot to take in there. But I would try to understand webpack and what it tries to do via babel as it will help you alot.

If you still need help let me know and I'll post webpack config although it will be a little project specific.

**Reply by se22as** [reply]
I am a little unclear after reading this page and other sites. I have created a react-native application which i am trying to add web support to. I am following the instructions on here https://medium.com/@aureliomerenda/create-a-native-web-app-with-react-native-web-419acac86b82

The last comment by @arelra is really clear on what is the issue and what needs to be done, but I do not entirely understand how to add the solution.

The above notes talk about adding stuff to webpack files, but any project created using the tools for web/native do not have any webpack files

creating an

-  web app with "create-react-app" results in a project with no webpack.config.js file, as the webpack stuff is wrapped in the react-scripts code
- creating a native react app with "npx react-native init ReactNativeWebApp" results in a project with no webpack.config.js file

I have a babel.config.js which currently only contains the following

```
module.exports = {
  presets: ['module:metro-react-native-babel-preset', '@babel/preset-flow'],
};

```
I have written React SSR apps from scratch, i.e. not using any CLI and therefore have added webpack to my project. However I am unclear how to do this for projects created with a CLI

**Reply by arelra** [reply]
Hi

@se22as see below for answers to CRA and webpack
@ZahraMirali see below for example webpack config

**Problem: RN dependencies are not being transpiled for web**
**Solution: Configure webpack to transpile RN dependencies for web**

I can see why this is tough as there are many layers to this issue i.e. RN, CRA, Webpack, Babel.

It also took me a long time to get the bottom of it, so keep persevering, it will make sense eventually.

To answer your specific questions:

**React Native Web Applications created with Create React App**

> web app with "create-react-app" results in a project with no webpack.config.js file, as the webpack stuff is wrapped in the react-scripts code

> However I am unclear how to do this for projects created with a CLI

create-react-app is awesome at getting quickly started but the downside (in this case) is it hides the configuration from you.

This will prevent you from making the changes required to transpile your react native dependencies.

You have two options with CRA-apps:

1. Eject your application so it removes CRA and gives you back all the configuration: [Ejecting](https://create-react-app.dev/docs/available-scripts#npm-run-eject)
2. Use https://github.com/arackaf/customize-cra or https://github.com/timarney/react-app-rewired to programatically modify your webpack config without ejecting

For this particular problem I would opt for ejecting completely but decide for yourself which you feel comfortable with

**Webpack config**

> The last comment by @arelra is really clear on what is the issue and what needs to be done, but I do not entirely understand how to add the solution.

So what we need to do is basically get webpack to pass all our react native dependencies through a preset which will transpile them so we can use them on web.

Assuming you have ejected using CRA here's the relevant parts of the webpack config:

(Note the commented lines )

```
// top level create path consts to use later
const repoIncludes = [
  paths.appSrc,
  path.resolve(paths.appPath, 'path from webpack.config.js to any shared component project'),
]
...
// comment out the following line if it isn't already as it prevents imports from outside the web project e.g.
// new ModuleScopePlugin(paths.appSrc, [paths.appPackageJson]),
...
module: {
  rules: [
  { eslint-loader },
  oneOf: [
    // in here we have lots of 'loaders' with 'test' regexes which webpack will use to match specific loaders against to do our transpiling for us
    // here's the loader for all our JS,TS, React etc, note the commented lines // <--
            {
              test: /\.(js|mjs|jsx|ts|tsx)$/,
              include: repoIncludes, // <-- picks up your monorepo src
              loader: require.resolve('babel-loader'),
              options: {
                customize: require.resolve(
                  'babel-preset-react-app/webpack-overrides'
                ),
                presets: [
                  ['babel-preset-react-app'],
                ],
                plugins: [
                  [
                    require.resolve('babel-plugin-react-native-web'), // <-- from the react-native-web install docs
                  ],
                  [
                    require.resolve('babel-plugin-named-asset-import'),
                    {
                      loaderMap: {
                        svg: {
                          ReactComponent:
                            '@svgr/webpack?-svgo,+titleProp,+ref![path]',
                        },
                      },
                    },
                  ],
                ],
                // This is a feature of `babel-loader` for webpack (not Babel itself).
                // It enables caching results in ./node_modules/.cache/babel-loader/
                // directory for faster rebuilds.
                cacheDirectory: true,
                // See #6846 for context on why cacheCompression is disabled
                cacheCompression: false,
                compact: isEnvProduction,
              },
            },
            // this is where we tell webpack to transpile our react-navigation* and react-native* dependencies
            {
              test: /(@?react-(navigation|native)).*\.(ts|js)x?$/,
              exclude: [/react-native-web/, /\.(native|ios|android)\.(ts|js)x?$/],
              loader: 'babel-loader',
              options: {
                presets: [
                  [
                    require.resolve('babel-preset-react-app/dependencies'), // <--- this fixes some module detection issues
                    { helpers: true },
                  ],
                  ['babel-preset-react-app'], // <--- this is the preset that will transpile everything for web (its the one used by CRA)
                ],
                cacheDirectory: true,
                // See #6846 for context on why cacheCompression is disabled
                cacheCompression: false,
                // Babel sourcemaps are needed for debugging into node_modules
                // code.  Without the options below, debuggers like VSCode
                // show incorrect code and set breakpoints on the wrong lines.
                sourceMaps: shouldUseSourceMap,
                inputSourceMap: shouldUseSourceMap,
              }
            },
            // now you just keep adding loaders for all of your react dependencies that need transpiling e.g. ui-kitten
            {
              test: /@?(ui-kitten|eva-design).*\.(ts|js)x?$/,
              loader: 'babel-loader',
              options: {
                presets: [
                  [
                    require.resolve('babel-preset-react-app/dependencies'),
                    { helpers: true },
                  ],
                  ['babel-preset-react-app'],
                ],
                cacheDirectory: true,
                // See #6846 for context on why cacheCompression is disabled
                cacheCompression: false,
                // Babel sourcemaps are needed for debugging into node_modules
                // code.  Without the options below, debuggers like VSCode
                // show incorrect code and set breakpoints on the wrong lines.
                sourceMaps: shouldUseSourceMap,
                inputSourceMap: shouldUseSourceMap,
              }
            },
            // other loaders
```

**react-native-web**

You should also follow the installation instructions for `react-native-web` and whatever config changes it requires.

**metro-react-native-babel-preset**

I've seen other answers use `metro-react-native-babel-preset` but that is the very heavy default RN preset meant for mobile JS engines. This will output CommonJS modules which means you won't be able to tree-shake your bundle. This will mean a larger build file and potentially slower start up time for your app.

**Future**

As time progresses RN dependencies are becoming more web friendly like [here](https://github.com/react-navigation/react-navigation/issues/8319) so eventually you won't need to configure custom loaders.

Also CRA might come up with workarounds to allow you to configure for a cross platform RN app.

**Summary**
 
For RN web+mobile apps, any errors due to not parsing JS from RN dependencies correctly are probably due to this underlying transpilation issue.

Hope that helps!

**Reply by se22as** [reply]
@arelra Thank you for your quick response. I have been playing with "customize-cra" and "react-app-rewired" and it is still not transpiling, or at least i am getting the same error.

I think my initial post was misleading, I apologise. I created my react native application using "npx react-native init MyTestApp" rather than using "create-react-app", I am guessing therefore "customize-cra" and "react-app-rewired" are not applicable as they are about extending a CRA app rather than a react-native app (created with "npx react-native init" rather than CRA").

**Reply by se22as** [reply]
@arelra I am still struggling greatly with this issue.  I simply can not get a very simple application with react-navigation to run on the web.

The latest thing i have tried is to not use react-scripts and do the webpack stuff myself. I am following the steps on this page this time : https://dev.to/ugglr/step-2-setting-up-react-with-react-native-web-30ba
I have done the following steps


1. create a reactnative app using "react-native init myProject"
2. installed all the dependencies needed for web (react-native-web, react-dom, @babel/core babel-loader @babel/preset-react @babel/preset-env, webpack webpack-cli webp
3. added the following to the package.json scripts
  "build-react": "webpack --mode production",
  "start-react": "webpack-dev-server --config ./webpack.config.js --mode development",
4. added webpack.config.js
5. added public/index.html (ensuring the html has a div with id="app")
6. added index.web.js which imports my main App.js component and renders it in the html

run the android with 

- npm run android

run the web with 

- npm run start-react

my babel.config.js is 

```
module.exports = {
  presets: ['module:metro-react-native-babel-preset'],
};
```

my webpack.config.js is

```
const path = require('path')
const HTMLWebpackPlugin = require('html-webpack-plugin')

const HTMLWebpackPluginConfig = new HTMLWebpackPlugin({
  template: path.resolve(__dirname, './public/index.html'),
  filename: 'index.html',
  inject: 'body',
})

module.exports = {
  entry: 
    path.join(__dirname, 'index.web.js'),
  
  output: {
    filename: 'bundle.js',
    path: path.join(__dirname, '/build'),
  },

  module: {
    rules: [
      // enable importing of images, e.g. import facebookImage from '../images/facebook.png';
      {
        test: /\.(png|jpe?g|gif)$/i,
        use: 'file-loader',
      },      

      // https://github.com/babel/babel/discussions/11694#discussioncomment-84474
      {
        test: /\.(js|mjs|jsx|ts|tsx)$/,
        loader: require.resolve('babel-loader'),
        options: {
          customize: require.resolve(
            'babel-preset-react-app/webpack-overrides'
          ),
          presets: [
            ['babel-preset-react-app'],
          ],
          plugins: [
            [
              require.resolve('babel-plugin-react-native-web'), // <-- from the react-native-web install docs
            ],
            [
              require.resolve('babel-plugin-named-asset-import'),
              {
                loaderMap: {
                  svg: {
                    ReactComponent:
                      '@svgr/webpack?-svgo,+titleProp,+ref![path]',
                  },
                },
              },
            ],
          ],
          // This is a feature of `babel-loader` for webpack (not Babel itself).
          // It enables caching results in ./node_modules/.cache/babel-loader/
          // directory for faster rebuilds.
          cacheDirectory: true,
          // See #6846 for context on why cacheCompression is disabled
          cacheCompression: false,
          //compact: isEnvProduction,
        },
      },
      // this is where we tell webpack to transpile our react-navigation* and react-native* dependencies
      {
        test: /(@?react-(navigation|native)).*\.(ts|js)x?$/,
        exclude: [/react-native-web/, /\.(native|ios|android)\.(ts|js)x?$/],
        loader: 'babel-loader',
        options: {
          presets: [
            [
              require.resolve('babel-preset-react-app/dependencies'), // <--- this fixes some module detection issues
              { helpers: true },
            ],
            ['babel-preset-react-app'], // <--- this is the preset that will transpile everything for web (its the one used by CRA)
          ],
          cacheDirectory: true,
          // See #6846 for context on why cacheCompression is disabled
          cacheCompression: false,
          // Babel sourcemaps are needed for debugging into node_modules
          // code.  Without the options below, debuggers like VSCode
          // show incorrect code and set breakpoints on the wrong lines.
          //sourceMaps: shouldUseSourceMap,
          //inputSourceMap: shouldUseSourceMap,
        }
      },
    ],
    
  },

  resolve: {
    alias: {
      'react-native$': 'react-native-web',
    },
  },

  plugins: [HTMLWebpackPluginConfig],

  devServer: {
    historyApiFallback: true,
    contentBase: './',
    hot: true,
  },
}
```

Running the web errors with 

```
Directions.js:1 Uncaught TypeError: Cannot read property 'Direction' of undefined
    at eval (Directions.js:1)
    at Object../node_modules/react-native-gesture-handler/Directions.js (bundle.js:2743)
```

Any idea whats causing this?

**Reply by arelra** [reply]
@se22as 

Googling throws up the following issue logs:

https://github.com/software-mansion/react-native-gesture-handler/issues/497
https://github.com/react-navigation/react-navigation/issues/5687

I would double check you have followed the RNGH installation steps correctly too:

https://reactnavigation.org/docs/getting-started/#installing-dependencies-into-a-bare-react-native-project

If you are still very stuck, you can add me as a collaborator on your repo and I might be able to diagnose better that way.

Also as your errors are now not related to this discussion, please discuss on another issue.

Thanks

**Reply by se22as** [reply]
I have finally got a resolution to my case using the information in this blog and Googling others.

I installed the following, to enable me to overwrite the settings in webpack and babel without having to do all the webpack stuff myself.

- react-native-scripts
- react-app-rewired
- customize-cra

I added the `webpack-overrides.js` file that `react-app-rewired` and `customize-cra` use to overwrite settings.  My file contents are 

```
const {
  override,
  addWebpackAlias,
  addBabelPlugins,
  addBabelPresets,
  addWebpackModuleRule,
} = require('customize-cra');

module.exports = override(
  addWebpackModuleRule({ test: /\.js$/, use: 'babel-loader', exclude: /node_modules\/(?!()\/).*/ }),
  ...addBabelPlugins('@babel/plugin-proposal-class-properties'),
  ...addBabelPresets('@babel/preset-env', '@babel/preset-react'),
);
```

To get applications with `react-navigation` working, add the following to the top of the main component (in my app this is `App.js`).

`import 'react-native-gesture-handler';`


Thank you very much for all your help in this discussion. It has helped me greatly.

**Reply by ankitLV** [reply]
Hey @se22as can you please give the detailed step by step solution, I am also facing the same issue with the react-navtigation, I also followed the same article "https://medium.com/@aureliomerenda/create-a-native-web-app-with-react-native-web-419acac86b82" , after that I am trying to setup react-navigation but I am getting the same error like you. It would be very helpful if you can provide me the step wise solution. Thanks in advance.

**Reply by AnmolShrma** [reply]
> I have finally got a resolution to my case using the information in this blog and Googling others.
> 
> I installed the following, to enable me to overwrite the settings in webpack and babel without having to do all the webpack stuff myself.
> 
> * react-native-scripts
> * react-app-rewired
> * customize-cra
> 
> I added the `webpack-overrides.js` file that `react-app-rewired` and `customize-cra` use to overwrite settings. My file contents are
> 
> ```
> const {
>   override,
>   addWebpackAlias,
>   addBabelPlugins,
>   addBabelPresets,
>   addWebpackModuleRule,
> } = require('customize-cra');
> 
> module.exports = override(
>   addWebpackModuleRule({ test: /\.js$/, use: 'babel-loader', exclude: /node_modules\/(?!()\/).*/ }),
>   ...addBabelPlugins('@babel/plugin-proposal-class-properties'),
>   ...addBabelPresets('@babel/preset-env', '@babel/preset-react'),
> );
> ```
> 
> To get applications with `react-navigation` working, add the following to the top of the main component (in my app this is `App.js`).
> 
> `import 'react-native-gesture-handler';`
> 
> Thank you very much for all your help in this discussion. It has helped me greatly.

@se22as Tried your solution but got the same error can you please let me know if I need to do the changes mentioned in this comment -> https://github.com/babel/babel/discussions/11694#discussioncomment-91186, too. If you can share the complete steps with the code changes that will be really helpful. Thank you

**Reply by lekzd** [reply]
@AnmolShrma I just copied react-scripts wenpack.config.js rules section with `test: /\.(js|mjs)$/` to my config-overrides.js and add 
'module:metro-react-native-babel-preset' to presets and '@babel/plugin-proposal-class-properties' to plugins

my final solution looks like:

config-override.js
```
const { override, addWebpackAlias, addBabelPlugins, addBabelPresets, addWebpackModuleRule } = require('customize-cra');

module.exports = override(
    addWebpackModuleRule({
        test: /\.(js|mjs)$/,
        exclude: /(@babel(?:\/|\\{1,2})runtime)/,
        loader: require.resolve('babel-loader'),
        options: {
            babelrc: false,
            configFile: false,
            compact: false,
            presets: [
                ['module:metro-react-native-babel-preset'],
                [
                    require.resolve('babel-preset-react-app/dependencies'),
                    { helpers: true },
                ],
            ],
            plugins: [
                ['@babel/plugin-proposal-class-properties']
            ],
            cacheDirectory: true,
            // See #6846 for context on why cacheCompression is disabled
            cacheCompression: false,
        },
    }),
);

```

**Comment by hanoisteve** [comment]
Can someone simply show the modified babel.config.js  that works? I have a react native app with react native web. It works on Android and iOS but I get an error for web.  I have read through the answers here but can we just clarify exactly the steps to solve this?  Is it just modifying the babel.config.js? If so can we confirm such a file that works? Or is some full ejection required?
![Screen Shot 2021-02-18 at 11 56 55](https://user-images.githubusercontent.com/50662937/108323598-72197b00-71f9-11eb-866f-8ecb24ae9630.png).

**Reply by jamesdo** [reply]
@hanoisteve, any chance you figured this out?
